### PR TITLE
Fix HashStable implementation on AllocId

### LINKED ISSUE #93470: ICE on match of const variables: destructure_const
I am getting an ICE on the following code.

The ICE is always reproducible in a clean build, but in an incremental build, I can bypass the ICE by removing the `foo` function, compiling, and then adding the `foo` function back again.

This is possibly related to #89690 or #83085 given the similar error message, but I am filing a new issue with my own reduced case because the stack is different.

### Code

Gist: https://gist.github.com/sffc/f09a3a468a82210a713dc9d8b408b46d

```Rust
#[derive(PartialEq, Eq)]
pub struct Key {
    path: &'static str,
}

pub const CONST_A: Key = Key {
    path: "time_zone/formats@1",
};

pub const CONST_B: Key = Key {
    path: "time_zone/formats@1",
};

fn foo(key: Key) -> Result<(), &'static str> {
    match key {
        CONST_B => Ok(()),
        _ => Err(""),
    }
}

fn bar(key: Key) -> Result<(), &'static str> {
    match key {
        CONST_A => Ok(()),
        _ => Err(""),
    }
}
```

### Meta

`rustc --version --verbose`:
```
$ rustc --version --verbose
rustc 1.58.1 (db9d1b20b 2022-01-20)
binary: rustc
commit-hash: db9d1b20bba1968c1ec1fc49616d4742c1725b4b
commit-date: 2022-01-20
host: x86_64-unknown-linux-gnu
release: 1.58.1
LLVM version: 13.0.0
```

### Error output

```
thread 'rustc' panicked at 'forcing query with already existing `DepNode`
- query-key: ParamEnvAnd { param_env: ParamEnv { caller_bounds: [], reveal: UserFacing }, value: Const { ty: Key, val: Value(ByRef { alloc: Allocation { bytes: [0, 0, 0, 0, 0, 0, 0, 0, 19, 0, 0, 0, 0, 0, 0, 0], relocations: Relocations(SortedMap { data: [(Size { raw: 0 }, alloc7)] }), init_mask: InitMask { blocks: [65535], len: Size { raw: 16 } }, align: Align { pow2: 3 }, mutability: Not, extra: () }, offset: Size { raw: 0 } }) } }
- dep-node: destructure_const(94ed374db4be2c96-604af2b3d231c21a)', /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/compiler/rustc_query_system/src/dep_graph/graph.rs:236:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.58.1 (db9d1b20b 2022-01-20) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [destructure_const] destructure constant
#1 [check_match] match-checking `bar`
end of query stack
error: could not compile `ice_destructure_const`
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'forcing query with already existing `DepNode`
- query-key: ParamEnvAnd { param_env: ParamEnv { caller_bounds: [], reveal: UserFacing }, value: Const { ty: Key, val: Value(ByRef { alloc: Allocation { bytes: [0, 0, 0, 0, 0, 0, 0, 0, 19, 0, 0, 0, 0, 0, 0, 0], relocations: Relocations(SortedMap { data: [(Size { raw: 0 }, alloc7)] }), init_mask: InitMask { blocks: [65535], len: Size { raw: 16 } }, align: Align { pow2: 3 }, mutability: Not, extra: () }, offset: Size { raw: 0 } }) } }
- dep-node: destructure_const(94ed374db4be2c96-604af2b3d231c21a)', /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/compiler/rustc_query_system/src/dep_graph/graph.rs:236:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /rustc/db9d1b20bba1968c1ec1fc49616d4742c1725b4b/library/core/src/panicking.rs:107:14
   2: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, rustc_middle::ty::ParamEnvAnd<&rustc_middle::ty::consts::Const>, rustc_middle::mir::query::DestructuredConst>
   3: rustc_data_structures::stack::ensure_sufficient_stack::<(rustc_middle::mir::query::DestructuredConst, rustc_query_system::dep_graph::graph::DepNodeIndex), rustc_query_system::query::plumbing::execute_job<rustc_query_impl::plumbing::QueryCtxt, rustc_middle::ty::ParamEnvAnd<&rustc_middle::ty::consts::Const>, rustc_middle::mir::query::DestructuredConst>::{closure#3}>
   4: rustc_query_system::query::plumbing::get_query::<rustc_query_impl::queries::destructure_const, rustc_query_impl::plumbing::QueryCtxt>
   5: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::destructure_const
   6: <rustc_mir_build::thir::pattern::const_to_pat::ConstToPat>::recur
   7: <rustc_infer::infer::InferCtxtBuilder>::enter::<rustc_middle::thir::Pat, <rustc_mir_build::thir::pattern::PatCtxt>::const_to_pat::{closure#0}>
   8: <rustc_mir_build::thir::pattern::PatCtxt>::lower_path
   9: <rustc_mir_build::thir::pattern::PatCtxt>::lower_pattern
  10: <rustc_mir_build::thir::pattern::check_match::MatchVisitor>::lower_pattern
  11: <alloc::vec::Vec<rustc_mir_build::thir::pattern::usefulness::MatchArm> as alloc::vec::spec_from_iter::SpecFromIter<rustc_mir_build::thir::pattern::usefulness::MatchArm, core::iter::adapters::map::Map<core::slice::iter::Iter<rustc_hir::hir::Arm>, <rustc_mir_build::thir::pattern::check_match::MatchVisitor>::check_match::{closure#0}>>>::from_iter
  12: <rustc_mir_build::thir::pattern::check_match::MatchVisitor as rustc_hir::intravisit::Visitor>::visit_expr
  13: <rustc_mir_build::thir::pattern::check_match::MatchVisitor as rustc_hir::intravisit::Visitor>::visit_expr
  14: rustc_mir_build::thir::pattern::check_match::check_match
  15: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, rustc_span::def_id::DefId, ()>
  16: rustc_data_structures::stack::ensure_sufficient_stack::<((), rustc_query_system::dep_graph::graph::DepNodeIndex), rustc_query_system::query::plumbing::execute_job<rustc_query_impl::plumbing::QueryCtxt, rustc_span::def_id::DefId, ()>::{closure#3}>
  17: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<rustc_span::def_id::DefId, ()>>
  18: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::check_match
  19: <rustc_middle::hir::map::Map>::par_body_owners::<rustc_interface::passes::analysis::{closure#1}::{closure#0}::{closure#0}::{closure#0}>
  20: <rustc_session::session::Session>::time::<(), rustc_interface::passes::analysis::{closure#1}>
  21: rustc_interface::passes::analysis
  22: <rustc_query_system::dep_graph::graph::DepGraph<rustc_middle::dep_graph::dep_node::DepKind>>::with_task::<rustc_middle::ty::context::TyCtxt, (), core::result::Result<(), rustc_errors::ErrorReported>>
  23: rustc_data_structures::stack::ensure_sufficient_stack::<(core::result::Result<(), rustc_errors::ErrorReported>, rustc_query_system::dep_graph::graph::DepNodeIndex), rustc_query_system::query::plumbing::execute_job<rustc_query_impl::plumbing::QueryCtxt, (), core::result::Result<(), rustc_errors::ErrorReported>>::{closure#3}>
  24: rustc_query_system::query::plumbing::try_execute_query::<rustc_query_impl::plumbing::QueryCtxt, rustc_query_system::query::caches::DefaultCache<(), core::result::Result<(), rustc_errors::ErrorReported>>>
  25: <rustc_query_impl::Queries as rustc_middle::ty::query::QueryEngine>::analysis
  26: <rustc_interface::passes::QueryContext>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}::{closure#3}, core::result::Result<(), rustc_errors::ErrorReported>>
  27: <rustc_interface::interface::Compiler>::enter::<rustc_driver::run_compiler::{closure#1}::{closure#2}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_errors::ErrorReported>>
  28: rustc_span::with_source_map::<core::result::Result<(), rustc_errors::ErrorReported>, rustc_interface::interface::create_compiler_and_run<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>::{closure#1}>
  29: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::setup_callbacks_and_run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_errors::ErrorReported>, rustc_driver::run_compiler::{closure#1}>::{closure#0}, core::result::Result<(), rustc_errors::ErrorReported>>::{closure#0}::{closure#0}, core::result::Result<(), rustc_errors::ErrorReported>>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.58.1 (db9d1b20b 2022-01-20) running on x86_64-unknown-linux-gnu

note: compiler flags: -C embed-bitcode=no -C debuginfo=2 -C incremental --crate-type lib

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
#0 [destructure_const] destructure constant
#1 [check_match] match-checking `bar`
#2 [analysis] running analysis passes on this crate
end of query stack
error: could not compile `ice_destructure_const`
```

</p>
</details>

### PR DESCRIPTION
**Pull request by compiler-errors**
(take 2:)

This PR removes the `HashStable` implementation from `AllocId`, and fixes the `HashStable` implementation on `ConstValue`, `ConstAlloc`, and `Allocation` so that their `Eq` and `HashStable` implementations agree (as required by `HashStable`).

Fixes #93470

r? @oli-obk

### CODE REVIEW

**Comment by compiler-errors** [comment]
Ok this definitely doesn't work :^)

**Comment by oli-obk** [comment]
The easy solution here is to remove the HashStable impl entirely and instead add a manual impl for `ConstAlloc` and `ConstValue` and other types that have an alloc id and require HashStable. Alternatively create a convenience wrapper type around AllodID with manual partialeq and stablehash impls and use it in those places.

**Comment by compiler-errors** [comment]
> The easy solution here is to remove the HashStable impl entirely and instead add a manual impl for ConstAlloc and ConstValue and other types that have an alloc id and require HashStable.

@oli-obk, I took an initial stab at that suggestion of yours. In order to implement `HashStable` correctly on those types mentioned above, I had to reimplement `Eq` (and friends) to agree with the `HashStable` implementation (that compares the allocation behind the `AllocId`). _However,_ this isn't a complete solution.. evident by the fact that I'm having trouble bootstrapping the compiler due to some const-eval issues:

```
error[E0080]: it is undefined behavior to use this value                                                                                                                                                                 
   --> /home/michael/.cargo/registry/src/github.com-1ecc6299db9ec823/getopts-0.2.21/src/lib.rs:927:1                                                                                                                     
    |                                                                                                                                                                                                                    
927 | fn is_arg(arg: &str) -> bool {                                                                                                                                                                                     
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed at .<deref>.<enum-variant(Some)>.0.<deref>: encountered (potentially part of) a pointer, but expected plain (non-pointer) bytes                                
    |                                                                                                                                                                                                                    
    = note: The rules on what exactly is undefined behavior aren't clear, so this check might be overzealous. Please open an issue on the rustc repository if you believe it should not be considered undefined behavior.
    = note: the raw bytes of the constant (size: 8, align: 8) {                                                                                                                                                          
                ╾──────alloc1253──────╼                         │ ╾──────╼                                                                                                                                               
            }                                                                                                                                                                                                            
                                                                                                                                                                                                                         
For more information about this error, try `rustc --explain E0080`.                                                                                                                                                      
```

I'm assuming some const eval code depends on that compare-alloc-by-id behavior, but I can't track it down. :shrug:

File: compiler/rustc_middle/src/ty/relate.rs [hunk 1]
```diff
@@ -634,7 +634,8 @@ fn check_const_value_eq<'tcx, R: TypeRelation<'tcx>>(
             ConstValue::Scalar(Scalar::Ptr(a_val, _a_size)),
             ConstValue::Scalar(Scalar::Ptr(b_val, _b_size)),
         ) => {
-            a_val == b_val
+            // FIXME(compiler-errors): This is highly suspicious
+            a_val.provenance.0 == b_val.provenance.0
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/relate.rs:638]: this is to preserve the old "equate AllocId by value" behavior.

**Comment by compiler-errors** [comment]
Alright, this definitely has to do with some code relying on `relocations` in `Allocation` being compared by id.

**Comment by compiler-errors** [comment]
Well this is a rough set of working changes. I can clean it up tomorrow...

**Comment by oli-obk** [comment]
Ah yes, sorry. I completely forgot about allocations themselves. Remove their HashStable impl and make it a free/inherent function. Then you can just keep calling it from ConstAlloc's impl

**Comment by oli-obk** [comment]
Oh, my browser didn't refresh. Ignore the previous comment. Reading the impl now

**Comment by oli-obk** [comment]
Looks like most of these manual impls will be gone when we get valtrees... so I am somewhat inclined to roll with the general idea. 

I was considering just swapping out the `AllocId` type (Scalar and Allocation are generic over it) so we could keep all derives but the one on whatever we use instead of `AllocId`.

**Comment by RalfJung** [comment]
I am a bit worried about considering allocations equal when their contents are equal (but their ID is not) -- this might make sense for some allocations, but certainly not for all of them (e.g., `static`). And moreover, the fact that the type system depends on this equality at all is a bug that we need to fix anyway (by using valtrees). So is it really worth introducing subtle and risky changes that won't fix the actual problem?

**Comment by oli-obk** [comment]
> I am a bit worried about considering allocations equal when their contents are equal (but their ID is not)

we're def not going to do that :) I'm more than worried about something like this. Hash/Eq should just be plain derives on `Allocation` and `AllocId`, but `HashStable` should not be implemented for either of them, and the surrounding types must handle that manually.

**Comment by compiler-errors** [comment]
> I was considering just swapping out the AllocId type (Scalar and Allocation are generic over it) so we could keep all derives but the one on whatever we use instead of AllocId.

@oli-obk: yeah, I considered that too, especially because `Tag` is generic all over the place, but (at least if I understand it correctly) then there's a deep-copy you need to perform when converting between AllocId and "AllocId that hashes/equates according to its contents"..

**Comment by oli-obk** [comment]
Yea, I agree it would kill perf, so let's keep doing what you were doing

File: compiler/rustc_middle/src/ty/impls_ty.rs [hunk 2]
```diff
@@ -132,35 +135,370 @@ where
...
+            ConstValue::Scalar(scalar) => {
+                std::mem::discriminant(scalar).hash_stable(hcx, hasher);
+                match scalar {
+                    Scalar::Int(int) => int.hash_stable(hcx, hasher),
+                    Scalar::Ptr(ptr, size) => {
+                        let tcx = tcx.expect("can't hash AllocIds during hir lowering");
+                        let (ptr, offset) = ptr.into_parts();
+                        tcx.get_global_alloc(ptr).hash_stable(hcx, hasher);
+                        offset.hash_stable(hcx, hasher);
+                        size.hash_stable(hcx, hasher);
```
Comment (b-naber) [inline compiler/rustc_middle/src/ty/impls_ty.rs:171]: Is it really necessary to hash `size`? This is the pointer size of the target, so identical for all pointers. Plus hashing is expensive.

Comment (compiler-errors) [reply compiler/rustc_middle/src/ty/impls_ty.rs:171]: Good point, but I'd rather implement a correct implementation of HashStable than an optimized one first. I'll leave a FIXME and get back to it.

**Comment by compiler-errors** [comment]
```
-                ╾─allocN─╼ ╾─allocN─╼                         │ ╾──╼╾──╼
+                ╾─allocN──╼ ╾─allocN─╼                         │ ╾──╼╾──╼
```

Literally don't know what's going on with these tests. I don't even understand what these diagrams are trying to convey :face_with_spiral_eyes:. 

I guess something is hashing one of these types that I fixed in the example and using them for diagnostics?

**Comment by oli-obk** [comment]
😄 these are normalized to hide the actual id, but that doesn't account for IDs with different number of digits.

Also you need to bless tests with `--target i686_unknown_linux_gnu` or whatever the appropriate 32 bit target is

**Comment by compiler-errors** [comment]
I'll try to take a stab at fixing these CI failures in a few days.

@rustbot author

**Comment by RalfJung** [comment]
> we're def not going to do that :)

Good. :) In that case I guess I don't understand what actually is being done here.

>  Hash/Eq should just be plain derives on Allocation and AllocId, but HashStable should not be implemented for either of them, and the surrounding types must handle that manually.

Should we have Hash/Eq at all on `Allocation`? Seems a bit odd to be comparing allocations, TBH.

**Comment by compiler-errors** [comment]
Well looks like CI is happy once I blessed the 32-bit tests, so

@rustbot ready

**Comment by oli-obk** [comment]
Is this PR still necessary with the realization in https://rust-lang.zulipchat.com/#narrow/stream/260443-project-const-generics/topic/valtrees/near/270830218 that we don't need to put alloc IDs into query keys?

**Comment by compiler-errors** [comment]
> we don't need to put alloc IDs into query keys?

my understanding is that this PR mostly addresses hashing the structs that are returned by various queries, not the ones used as keys

**Comment by oli-obk** [comment]
Right, but most of these will become obsolete with valtrees.

The ones that don't, should actually be using alloc IDs by their identity, as otherwise static items may get the deduplication, too, and that is bad. We have existing bugs around this for mutable references in static mut (and probably interior mut ones in plain statics)

**Comment by compiler-errors** [comment]
I am not opposed to closing this PR in favor of waiting for valtrees to be the proper fix of representing stable const values returned by queries. Lemme know if that's what you mean and I can close this out.

**Comment by oli-obk** [comment]
I do. Closing then, thanks for trying it out!

**Pull request closed without merging**